mtec/mtec.cpp

```diff
@@ -0,0 +1,155 @@
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
+*
+* Copyright 2013 - 2021, nymea GmbH
+* Contact: contact@nymea.io
+*
+* This file is part of nymea.
+* This project including source code and documentation is protected by
+* copyright law, and remains the property of nymea GmbH. All rights, including
+* reproduction, publication, editing and translation, are reserved. The use of
+* this project is subject to the terms of a license agreement to be concluded
+* with nymea GmbH in accordance with the terms of use of nymea GmbH, available
+* under https://nymea.io/license
+*
+* GNU Lesser General Public License Usage
+* Alternatively, this project may be redistributed and/or modified under the
+* terms of the GNU Lesser General Public License as published by the Free
+* Software Foundation; version 3. This project is distributed in the hope that
+* it will be useful, but WITHOUT ANY WARRANTY; without even the implied
+* warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+* Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public License
+* along with this project. If not, see <https://www.gnu.org/licenses/>.
+*
+* For any further details and any questions please contact us under
+* contact@nymea.io or see our FAQ/Licensing Information on
+* https://nymea.io/license/faq
+*
+* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+#include "mtec.h"
+#include "extern-plugininfo.h"
+#include "mtechelpers.h"
+
+#include <cstring>
+
+MTec::MTec(const QHostAddress &address, QObject *parent) :
+    QObject(parent),
+    m_hostAddress(address)
+{
+    m_modbusMaster = new ModbusTCPMaster(address, 502, this);
+
+    if (m_modbusMaster) {
+        qCDebug(dcMTec()) << "created ModbusTCPMaster";
+        m_modbusMaster->connectDevice();
+
+        connect(m_modbusMaster, &ModbusTCPMaster::receivedHoldingRegister, this, &MTec::onReceivedHoldingRegister);
+        connect(m_modbusMaster, &ModbusTCPMaster::readRequestError, this, &MTec::onModbusError);
+        connect(m_modbusMaster, &ModbusTCPMaster::writeRequestError, this, &MTec::onModbusError);
+    }
+}
+
+MTec::~MTec()
+{
+    if (m_modbusMaster) {
+        delete m_modbusMaster;
+    }
+}
+
+void MTec::onModbusError()
+{
+    qCDebug(dcMTec()) << "MTec: Received modbus error";
```

Comment:
Maybe inform about which modbus error occured, i saw that is currently missing in the ModbusTcpMaster object, but we should implement it. Otherwise you just know, there is an error, but not which.

Also, use `qCWarning` so it will be printed to the logs even if the MTec logging category is not enabled. This is important to find issues in production setups.

Reply:
Changed to qCWarning.